### ISSUE TITLE
Preprocessor does not use the load method of the BioFile class

### ISSUE THREAD

**Issue by 183amir** [bioidiap/bob.bio.base#29]
The databases provide a load method in its `File` class but this is not used in the preprocessor when data is read.
`read_original_data`: https://github.com/bioidiap/bob.bio.base/blob/master/bob/bio/base/preprocessor/Preprocessor.py#L87

Some databases may contain more than one sample in one file (like videos and audios with two channles); while I understand that this is handled in `bob.bio.video` for video files, it is not clear how this can be handled for audio files with two channels in them.

If the preprocessor was calling the `load` method of the `File` (`BioFile`) class, we could use logical paths for `File.path` instead of the actual path and handle this in the `load` method. For example, 
`File.path` would be `origpath_A` or `origpath_B` depending on the channel and then the `load` method would return only channel A or B depending on the logical path that was requested.

**Comment by siebenkopf** [comment]
The ``File.load`` function (which AFAIR only exists in the ``bob.db.verification.utils.File`` class, not in ``bob.bio.base.database.File``) is not used to load the original data -- at least it was not the intention of this function. It rather exists for compatibility purposes, as the File in the old ``xbob.db`` databases contained such a function. In this implementation, it was supposed to read and write any kind of data -- usually using HDF5.

However, if the channels are stored in the same physical file, you could implement a specific audio preprocessor that returns the one or the other channel based on the preprocessor configuration.

In fact, if you have a protocol that requires both types of channels, you should have two ``File`` instances for the same physical file, one for each channel. Then, you can use the ``File.make_path`` function to add the channel name to each file name, i.e., when a unique filename is required, and read both channels (and possibly let the preprocessor decide, which data to use) when reading the original data. With that I mean:

Add an ``add_channel`` parameter to the ``File.make_path`` function:
```
def make_path(self, ..., add_channel = True):
  ...
  if make_channel:
    filename = self.path + "_" + self.channel
  else:
    filename = self.path
  ...
```

and call this function using ``add_channel=False`` in the ``Database.original_file_name`` function. This will allow to read the same physical file for two different ``File`` objects. Now, to tell the preprocessor, which channel to use, you can return the channel as an annotation of the file (inside ``Database.annotation``), such as:

```
def annotation (self, file):
  return {"channel" : file.channel}
```

and let the preprocessor decide, what to do with this information (usually, load only this channel).
I have implemented a similar solution for image databases, where two ore more people in the same physical image were annotated. There, I used the ``client_id`` instead of the channel in order to create a unique filename in the ``File.make_path`` function.

Anyways, this was only my solution to this specific problem. If you find a better solution (which might include revitalizing the ``File.load`` function -- I would rather call it ``File.read`` to be consistent with the usage of the terms load and read throughout ``bob.bio``) I would be happy to review a pull request.

**Comment by 183amir** [comment]
How about this: https://github.com/bioidiap/bob.bio.base/commit/11a7e2292f46229d39da2dd0161eb7073788f4a4

**Comment by siebenkopf** [comment]
I don't think this is a good idea. Loading of data should be done by the preprocessor, as only it knows how it wants to load the data. Your solution might work for any preprocessor that is able to process numpy arrays, but what about Bob-external preprocessors that use a different way of loading data, e.g.: 
https://github.com/bioidiap/bob.bio.csu/blob/master/bob/bio/csu/preprocessor/LDAIR.py#L144
or even more tricky, when the preprocessor does not understand Python at all (such as in extensions using external/commercial software)?

I would prefer to go the way of annotations. However, I don't exactly know your problem, so it is difficult for me to come up with a solution. Maybe you can be more precise in what you want to achieve, and why it does not work with the current implementation...

**Comment by anjos** [comment]
We have been through similar discussions in the past. I think the real issue here is that we're talking about "files" while we should be discussing in terms of "samples". Sometimes, these samples are available on disk and can be cleanly loaded. Sometimes they simply can't.

I understand this is why Amir is suggesting loading data from a ``bob.db`` interface should be handled by the interface itself and not by a specific framework. In specific cases, these frameworks (like ``bob.bio.base``) cannot handle this efficiently unless a lot of information is provided by the interface. As an example, imagine what you would have to do to properly handle loading data from a SQL database.

Let me gather the use-cases from the discussions above:

1. Samples in a database may not be stored as 1 sample per file (e.g. samples stored in a SQL database or multiple samples stored in a single file)
2. Samples in a database may have to be loaded differently depending on the sample (databases with a mixture of PNG and JPEG images for instance)
3. Data may be fed into an external binary that can only handle files (maybe pipes as well)

Here is one solution that may work for all the cases above:

a. The database designer defines the concept of a ``Sample`` (N.B.: for as long as 1 file = 1 sample, the ``Sample`` can as well as be our existing ``File``, what is convenient)
b. The database designer defines a ``load()`` method that can load the said sample with all its weird specifications, from an ``original_directory``.
c. The database designer defines a ``make_path()`` method that can return a unique path for a given ``Sample`` in the database interface.

Then, we do the following (in pseudo-code, details omitted):

```python
# Specific Preprocessor call implementation that can handle numpy arrays
def __call__(self, sample):
    data = sample.load(original_directory) #don't know where sample comes from
    annotations = sample.annotations()
    annotations = sample.annotations()
    # do your thing, setup ``result``
    return result
```

The temporary ``result`` may then be saved into a file using the Sample's ``make_path()``, which should return a unique path for each *sample* in the database.

For case ``c.`` above, we can do like this, on the specific preprocessor implementation (which is anyways the only entity that should know how to run the preprocessor properly):

```python
# Specific Preprocessor call implementation that can only handle files
def __call__(self, sample):
    use_path = sample.make_path(original_directory)
    if not os.path.exists(use_path): #database is weird, save to temp file
        use_path = tempfile.NamedTemporaryFile() #auto-deletes after file is closed
        data = sample.load() #don't know where sample comes from
        # save ``data`` on ``use_path`` handle
    # do your thing, setup ``result`` from ``use_path``
    return result
```

If you wish, you may abstract this on a base class "FilePreprocessor".

Does that clarify?

**Comment by siebenkopf** [comment]
So far, I have used the ``File`` instance as what you would call a ``Sample``. In most cases, I had 1 file = 1 sample, but for video databases, this is clearly not the case. There, we had one sample per video, which in case of the YouTube dataset is split over several image files.

Although I kind-of like the approach that you are proposing, so far I tried to keep the interface to the tools (preprocessor, extractor, algorithm) to be as simple as possible, so that it is easy to implement them independent of any (specific) Bob database. Instead, all parameters to functions of the tools depend only on simple data types -- such as strings for file names. In your approach you violate that concept by introducing a complex data structure as parameter to a tool (a preprocessor). This might be difficult to understand for people trying to implement a new preprocessor. That's why I am thinking of different approaches in order to keep the interface as simple as possible -- while possibly making the backend more complicated.

In fact, the actual place to implement a custom solution to ``read`` the data is the ``read_original_data`` function, and not inside the ``__call__`` function. The former function is particularly designed to handle cases like that, where you have more difficult structures. So far, this function receives an image filename and load the original data using ``bob.io.base.load``. As this function has a proper default implementation in ``bob.bio.base.Preprocessor``:
https://github.com/bioidiap/bob.bio.base/blob/master/bob/bio/base/preprocessor/Preprocessor.py#L87
there I could see a modification of the default parameter to make sense. In there you might want to use an instance of a ``File`` to load the data. But, please do not change the interface of functions that everyone needs to implement.

So, at the end, this modification would be similar to what @183amir proposed. However, I wouldn't want the modification in the ``tools.preprocessor`` function (since this is global to every preprocessor), but in the ``Preprocessor`` base class, so that custom solutions are still possible.

So, what do you think about this?
